src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -120,6 +123,33 @@ protected override void OnEventWritten(EventWrittenEventArgs eventData)
     private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
         new() { { "EventCounterIntervalSec", options.RefreshIntervalSecs.ToString() } };
 
+    /// <summary>
+    /// If instrument name is too long, abbreviates event source name.
+    /// E.g. instrument for `Microsoft-AspNetCore-Server-Kestrel`, `tls-handshakes-per-second`
+    /// would be too long (64 chars), so it's shortened to `ec.Microsoft-AspNetCore-Server-Kestre.tls-handshakes-per-second`.
+    /// 
+    /// If there is no room for event source name, returns `ec.{event name}` and 
+    /// if it's still too long, it will be validated and ignored later in the pipeline.
+    /// </summary>
+    private static string GetInstrumentName(string sourceName, string eventName)
+    {
+        int totalLength = Prefix.Length + 1 + sourceName.Length + 1 + eventName.Length;
+        if (totalLength <= MaxInstrumentNameLength)
+        {
+            return string.Concat(Prefix, ".", sourceName, ".", eventName);
+        }
+
+        var maxEventSourceLength = MaxInstrumentNameLength - Prefix.Length - 1 - eventName.Length;
+        if (maxEventSourceLength < 2)
+        {
+            // event name is too long, there is not enough space for sourceName.
+            // let ec.<eventName> flow to metrics SDK and it will suppress it if needed.
+            return string.Concat(Prefix, ".", eventName);
+        }
+
+        return string.Concat(Prefix, ".", sourceName.Substring(0, maxEventSourceLength - 1), ".", eventName);
```

Comment:
nit: Probably more readable to have `var maxEventSourceLength = MaxInstrumentNameLength - Prefix.Length - 2 - eventName.Length;

and use `sourceName.Substring(0, maxEventSourceLength)`